File: python/exercises/code-smells/divergent-change/src/customer_service.py
"""Customer service module with divergent change code smell."""

import hashlib
import re
from datetime import datetime, timedelta
from typing import Any


class CustomerService:
    """Service managing all customer-related operations (with divergent change issues)."""

    def __init__(self) -> None:
        """Initialize customer service with empty storage."""
        self._customers: dict[int, dict[str, Any]] = {}
        self._login_attempts: dict[str, list[datetime]] = {}
        self._marketing_preferences: dict[int, dict[str, Any]] = {}
        self._order_history: dict[int, dict[int, dict[str, Any]]] = {}
        self._addresses: dict[int, dict[int, dict[str, Any]]] = {}

    def register_customer(
        self, email: str, password: str, first_name: str, last_name: str
    ) -> int:
        """Register a new customer in the system.

        Args:
            email: Customer's email address
            password: Customer's password (min 8 characters)
            first_name: Customer's first name
            last_name: Customer's last name

        Returns:
            The ID of the newly registered customer

        Raises:
            ValueError: If input validation fails
        """
        if not self._is_valid_email(email):
            raise ValueError("Invalid email format")

        if len(password) < 8:
            raise ValueError("Password must be at least 8 characters")

        if not first_name or not last_name:
            raise ValueError("First name and last name are required")

        for customer in self._customers.values():
            if customer["email"] == email:
                raise ValueError("Customer with this email already exists")

        customer_id = len(self._customers) + 1
        self._customers[customer_id] = {
            "id": customer_id,
            "email": email,
            "password": self._hash_password(password),
            "first_name": first_name,
            "last_name": last_name,
            "phone": None,
            "status": "active",
            "created_at": datetime.now(),
            "last_login": None,
        }

        self._marketing_preferences[customer_id] = {
            "email_marketing": True,
            "sms_marketing": False,
            "push_notifications": True,
            "preferred_channels": ["email"],
        }

        return customer_id

    def authenticate_customer(self, email: str, password: str) -> int | None:
        """Authenticate a customer by email and password.

        Args:
            email: Customer's email address
            password: Customer's password

        Returns:
            Customer ID if authentication successful, None otherwise

        Raises:
            RuntimeError: If account is locked due to failed attempts
        """
        customer = self._find_customer_by_email(email)

        if not customer:
            self._record_failed_login_attempt(email)
            return None

        if self._is_account_locked(customer["id"]):
            raise RuntimeError(
                "Account is temporarily locked due to too many failed attempts"
            )

        if not self._verify_password(password, customer["password"]):
            self._record_failed_login_attempt(email)
            return None

        self._clear_failed_login_attempts(customer["id"])
        self._update_last_login(customer["id"])

        return customer["id"]

    def update_contact_information(
        self, customer_id: int, first_name: str, last_name: str, phone: str
    ) -> None:
        """Update customer's contact information.

        Args:
            customer_id: Customer's ID
            first_name: Updated first name
            last_name: Updated last name
            phone: Updated phone number

        Raises:
            ValueError: If customer not found or validation fails
        """
        if customer_id not in self._customers:
            raise ValueError("Customer not found")

        if not first_name or not last_name:
            raise ValueError("First name and last name are required")

        if phone and not self._is_valid_phone(phone):
            raise ValueError("Invalid phone number format")

        self._customers[customer_id]["first_name"] = first_name
        self._customers[customer_id]["last_name"] = last_name
        self._customers[customer_id]["phone"] = phone

    def add_customer_address(
        self,
        customer_id: int,
        street: str,
        city: str,
        zip_code: str,
        country: str,
        is_default: bool = False,
    ) -> int:
        """Add a new address for a customer.

        Args:
            customer_id: Customer's ID
            street: Street address
            city: City name
            zip_code: ZIP/postal code
            country: Country name
            is_default: Whether this should be the default address

        Returns:
            The ID of the newly created address

        Raises:
            ValueError: If customer not found or validation fails
        """
        if customer_id not in self._customers:
            raise ValueError("Customer not found")

        if not all([street, city, zip_code, country]):
            raise ValueError("All address fields are required")

        if customer_id not in self._addresses:
            self._addresses[customer_id] = {}

        address_id = len(self._addresses[customer_id]) + 1

        if is_default:
            for address in self._addresses[customer_id].values():
                address["is_default"] = False

        self._addresses[customer_id][address_id] = {
            "id": address_id,
            "street": street,
            "city": city,
            "zip_code": zip_code,
            "country": country,
            "is_default": is_default or len(self._addresses[customer_id]) == 0,
        }

        return address_id

    def update_marketing_preferences(
        self,
        customer_id: int,
        email_marketing: bool,
        sms_marketing: bool,
        push_notifications: bool,
        preferred_channels: list[str],
    ) -> None:
        """Update customer's marketing preferences.

        Args:
            customer_id: Customer's ID
            email_marketing: Allow email marketing
            sms_marketing: Allow SMS marketing
            push_notifications: Allow push notifications
            preferred_channels: List of preferred marketing channels

        Raises:
            ValueError: If customer not found or invalid channels provided
        """
        if customer_id not in self._customers:
            raise ValueError("Customer not found")

        valid_channels = ["email", "sms", "push", "mail"]
        for channel in preferred_channels:
            if channel not in valid_channels:
                raise ValueError(f"Invalid marketing channel: {channel}")

        self._marketing_preferences[customer_id] = {
            "email_marketing": email_marketing,
            "sms_marketing": sms_marketing,
            "push_notifications": push_notifications,
            "preferred_channels": preferred_channels,
        }

    def send_marketing_campaign(
        self, customer_id: int, subject: str, content: str, channel: str
    ) -> bool:
        """Send marketing campaign to customer.

        Args:
            customer_id: Customer's ID
            subject: Campaign subject
            content: Campaign content
            channel: Marketing channel to use

        Returns:
            True if campaign was sent, False if customer opted out

        Raises:
            ValueError: If customer not found or invalid channel
        """
        if customer_id not in self._customers:
            raise ValueError("Customer not found")

        if customer_id not in self._marketing_preferences:
            return False

        preferences = self._marketing_preferences[customer_id]

        if channel == "email":
            if (
                not preferences["email_marketing"]
                or "email" not in preferences["preferred_channels"]
            ):
                return False
        elif channel == "sms":
            if (
                not preferences["sms_marketing"]
                or "sms" not in preferences["preferred_channels"]
            ):
                return False
        elif channel == "push":
            if (
                not preferences["push_notifications"]
                or "push" not in preferences["preferred_channels"]
            ):
                return False
        else:
            raise ValueError(f"Unsupported marketing channel: {channel}")

        return True

    def record_purchase(
        self, customer_id: int, items: list[dict[str, Any]], total_amount: float
    ) -> int:
        """Record a customer purchase.

        Args:
            customer_id: Customer's ID
            items: List of purchased items
            total_amount: Total purchase amount

        Returns:
            The ID of the created order

        Raises:
            ValueError: If customer not found or validation fails
        """
        if customer_id not in self._customers:
            raise ValueError("Customer not found")

        if not items:
            raise ValueError("Purchase must contain at least one item")

        if total_amount <= 0:
            raise ValueError("Total amount must be positive")

        if customer_id not in self._order_history:
            self._order_history[customer_id] = {}

        order_id = len(self._order_history[customer_id]) + 1
        self._order_history[customer_id][order_id] = {
            "id": order_id,
            "items": items,
            "total_amount": total_amount,
            "order_date": datetime.now(),
            "status": "completed",
        }

        return order_id

    def get_customer_spending_history(self, customer_id: int) -> list[dict[str, Any]]:
        """Get customer's spending history.

        Args:
            customer_id: Customer's ID

        Returns:
            List of spending history entries, sorted by date (newest first)

        Raises:
            ValueError: If customer not found
        """
        if customer_id not in self._customers:
            raise ValueError("Customer not found")

        if customer_id not in self._order_history:
            return []

        history = []
        for order in self._order_history[customer_id].values():
            history.append(
                {
                    "order_id": order["id"],
                    "amount": order["total_amount"],
                    "date": order["order_date"],
                    "item_count": len(order["items"]),
                }
            )

        history.sort(key=lambda x: x["date"], reverse=True)
        return history

    def calculate_customer_lifetime_value(self, customer_id: int) -> float:
        """Calculate customer's lifetime value.

        Args:
            customer_id: Customer's ID

        Returns:
            Total amount spent by customer

        Raises:
            ValueError: If customer not found
        """
        if customer_id not in self._customers:
            raise ValueError("Customer not found")

        if customer_id not in self._order_history:
            return 0.0

        total_spent = 0.0
        for order in self._order_history[customer_id].values():
            total_spent += order["total_amount"]

        return total_spent

    def get_customer_profile(self, customer_id: int) -> dict[str, Any]:
        """Get complete customer profile.

        Args:
            customer_id: Customer's ID

        Returns:
            Dictionary containing all customer information

        Raises:
            ValueError: If customer not found
        """
        if customer_id not in self._customers:
            raise ValueError("Customer not found")

        customer = self._customers[customer_id]

        return {
            "personal": {
                "id": customer["id"],
                "first_name": customer["first_name"],
                "last_name": customer["last_name"],
                "email": customer["email"],
                "phone": customer["phone"],
                "status": customer["status"],
            },
            "addresses": self._addresses.get(customer_id, {}),
            "marketing": self._marketing_preferences.get(customer_id),
            "order_history": self.get_customer_spending_history(customer_id),
            "lifetime_value": self.calculate_customer_lifetime_value(customer_id),
        }

    def _find_customer_by_email(self, email: str) -> dict[str, Any] | None:
        """Find customer by email address."""
        for customer in self._customers.values():
            if customer["email"] == email:
                return customer
        return None

    def _record_failed_login_attempt(self, email: str) -> None:
        """Record a failed login attempt for an email."""
        if email not in self._login_attempts:
            self._login_attempts[email] = []
        self._login_attempts[email].append(datetime.now())

    def _is_account_locked(self, customer_id: int) -> bool:
        """Check if account is locked due to failed login attempts."""
        customer = self._customers[customer_id]
        email = customer["email"]

        if email not in self._login_attempts:
            return False

        recent_attempts = [
            attempt
            for attempt in self._login_attempts[email]
            if attempt > datetime.now() - timedelta(minutes=15)
        ]

        return len(recent_attempts) >= 3

    def _clear_failed_login_attempts(self, customer_id: int) -> None:
        """Clear failed login attempts for a customer."""
        customer = self._customers[customer_id]
        email = customer["email"]
        if email in self._login_attempts:
            del self._login_attempts[email]

    def _update_last_login(self, customer_id: int) -> None:
        """Update customer's last login timestamp."""
        self._customers[customer_id]["last_login"] = datetime.now()

    def _is_valid_email(self, email: str) -> bool:
        """Validate email format."""
        pattern = r"^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$"
        return re.match(pattern, email) is not None

    def _is_valid_phone(self, phone: str) -> bool:
        """Validate phone number format."""
        pattern = r"^\+?[1-9]\d{1,14}$"
        return re.match(pattern, phone) is not None

    def _hash_password(self, password: str) -> str:
        """Hash password using SHA-256."""
        return hashlib.sha256(password.encode()).hexdigest()

    def _verify_password(self, password: str, hashed: str) -> bool:
        """Verify password against hash."""
        return self._hash_password(password) == hashed

File: typescript/exercises/code-smells/divergent-change/src/CustomerService.ts
export type CustomerStatus = 'active' | 'inactive' | 'suspended';
export type MarketingChannel = 'email' | 'sms' | 'push' | 'mail';
export type OrderStatus = 'pending' | 'completed' | 'cancelled';

export interface Customer {
  readonly id: number;
  readonly email: string;
  readonly password: string;
  readonly firstName: string;
  readonly lastName: string;
  readonly status: CustomerStatus;
  readonly createdAt: Date;
  readonly lastLogin: Date | null;
  readonly phone?: string;
}

export interface Address {
  readonly id: number;
  readonly street: string;
  readonly city: string;
  readonly zipCode: string;
  readonly country: string;
  readonly isDefault: boolean;
}

export interface MarketingPreferences {
  readonly emailMarketing: boolean;
  readonly smsMarketing: boolean;
  readonly pushNotifications: boolean;
  readonly preferredChannels: readonly MarketingChannel[];
}

export interface OrderItem {
  readonly name: string;
  readonly price: number;
  readonly quantity: number;
}

export interface Order {
  readonly id: number;
  readonly items: readonly OrderItem[];
  readonly totalAmount: number;
  readonly orderDate: Date;
  readonly status: OrderStatus;
}

export interface SpendingHistoryEntry {
  readonly orderId: number;
  readonly amount: number;
  readonly date: Date;
  readonly itemCount: number;
}

export interface CustomerProfile {
  readonly personal: {
    readonly id: number;
    readonly firstName: string;
    readonly lastName: string;
    readonly email: string;
    readonly phone: string | null;
    readonly status: CustomerStatus;
  };
  readonly addresses: readonly Address[];
  readonly marketing: MarketingPreferences | null;
  readonly orderHistory: readonly SpendingHistoryEntry[];
  readonly lifetimeValue: number;
}

export class CustomerService {
  private customers: Map<number, Customer> = new Map();
  private loginAttempts: Map<string, Date[]> = new Map();
  private marketingPreferences: Map<number, MarketingPreferences> = new Map();
  private orderHistory: Map<number, Order[]> = new Map();
  private addresses: Map<number, Address[]> = new Map();
  private nextCustomerId: number = 1;

  public registerCustomer(
    email: string,
    password: string,
    firstName: string,
    lastName: string
  ): number {
    if (!this.isValidEmail(email)) {
      throw new Error('Invalid email format');
    }

    if (password.length < 8) {
      throw new Error('Password must be at least 8 characters');
    }

    if (!firstName.trim() || !lastName.trim()) {
      throw new Error('First name and last name are required');
    }

    if (this.findCustomerByEmail(email)) {
      throw new Error('Customer with this email already exists');
    }

    const customerId = this.nextCustomerId++;
    const customer: Customer = {
      id: customerId,
      email,
      password: this.hashPassword(password),
      firstName,
      lastName,
      status: 'active',
      createdAt: new Date(),
      lastLogin: null,
    };

    this.customers.set(customerId, customer);

    const defaultPreferences: MarketingPreferences = {
      emailMarketing: true,
      smsMarketing: false,
      pushNotifications: true,
      preferredChannels: ['email'],
    };
    this.marketingPreferences.set(customerId, defaultPreferences);

    return customerId;
  }

  public authenticateCustomer(email: string, password: string): number | null {
    const customer = this.findCustomerByEmail(email);

    if (!customer) {
      this.recordFailedLoginAttempt(email);
      return null;
    }

    if (this.isAccountLocked(customer.id)) {
      throw new Error(
        'Account is temporarily locked due to too many failed attempts'
      );
    }

    if (!this.verifyPassword(password, customer.password)) {
      this.recordFailedLoginAttempt(email);
      return null;
    }

    this.clearFailedLoginAttempts(customer.id);
    this.updateLastLogin(customer.id);

    return customer.id;
  }

  public updateContactInformation(
    customerId: number,
    firstName: string,
    lastName: string,
    phone: string
  ): void {
    const customer = this.customers.get(customerId);
    if (!customer) {
      throw new Error('Customer not found');
    }

    if (!firstName.trim() || !lastName.trim()) {
      throw new Error('First name and last name are required');
    }

    if (phone && !this.isValidPhone(phone)) {
      throw new Error('Invalid phone number format');
    }

    const updatedCustomer: Customer = {
      ...customer,
      firstName,
      lastName,
      phone,
    };

    this.customers.set(customerId, updatedCustomer);
  }

  public addCustomerAddress(
    customerId: number,
    street: string,
    city: string,
    zipCode: string,
    country: string,
    isDefault: boolean = false
  ): number {
    if (!this.customers.has(customerId)) {
      throw new Error('Customer not found');
    }

    if (!street.trim() || !city.trim() || !zipCode.trim() || !country.trim()) {
      throw new Error('All address fields are required');
    }

    const customerAddresses = this.addresses.get(customerId) || [];
    const addressId = customerAddresses.length + 1;

    if (isDefault) {
      // Remove default flag from existing addresses
      customerAddresses.forEach((addr, index) => {
        customerAddresses[index] = { ...addr, isDefault: false };
      });
    }

    const newAddress: Address = {
      id: addressId,
      street,
      city,
      zipCode,
      country,
      isDefault: isDefault || customerAddresses.length === 0,
    };

    const updatedAddresses = [...customerAddresses, newAddress];
    this.addresses.set(customerId, updatedAddresses);

    return addressId;
  }

  public updateMarketingPreferences(
    customerId: number,
    emailMarketing: boolean,
    smsMarketing: boolean,
    pushNotifications: boolean,
    preferredChannels: MarketingChannel[]
  ): void {
    if (!this.customers.has(customerId)) {
      throw new Error('Customer not found');
    }

    const validChannels: MarketingChannel[] = ['email', 'sms', 'push', 'mail'];
    for (const channel of preferredChannels) {
      if (!validChannels.includes(channel)) {
        throw new Error(`Invalid marketing channel: ${channel}`);
      }
    }

    const preferences: MarketingPreferences = {
      emailMarketing,
      smsMarketing,
      pushNotifications,
      preferredChannels,
    };

    this.marketingPreferences.set(customerId, preferences);
  }

  public sendMarketingCampaign(
    customerId: number,
    _subject: string,
    _content: string,
    channel: MarketingChannel
  ): boolean {
    if (!this.customers.has(customerId)) {
      throw new Error('Customer not found');
    }

    const preferences = this.marketingPreferences.get(customerId);
    if (!preferences) {
      return false;
    }

    switch (channel) {
      case 'email':
        return (
          preferences.emailMarketing &&
          preferences.preferredChannels.includes('email')
        );
      case 'sms':
        return (
          preferences.smsMarketing &&
          preferences.preferredChannels.includes('sms')
        );
      case 'push':
        return (
          preferences.pushNotifications &&
          preferences.preferredChannels.includes('push')
        );
      case 'mail':
        return preferences.preferredChannels.includes('mail');
      default:
        throw new Error(`Unsupported marketing channel: ${channel}`);
    }
  }

  public recordPurchase(
    customerId: number,
    items: OrderItem[],
    totalAmount: number
  ): number {
    if (!this.customers.has(customerId)) {
      throw new Error('Customer not found');
    }

    if (items.length === 0) {
      throw new Error('Purchase must contain at least one item');
    }

    if (totalAmount <= 0) {
      throw new Error('Total amount must be positive');
    }

    const customerOrders = this.orderHistory.get(customerId) || [];
    const orderId = customerOrders.length + 1;

    const order: Order = {
      id: orderId,
      items,
      totalAmount,
      orderDate: new Date(),
      status: 'completed',
    };

    const updatedOrders = [...customerOrders, order];
    this.orderHistory.set(customerId, updatedOrders);

    return orderId;
  }

  public getCustomerSpendingHistory(
    customerId: number
  ): SpendingHistoryEntry[] {
    if (!this.customers.has(customerId)) {
      throw new Error('Customer not found');
    }

    const orders = this.orderHistory.get(customerId) || [];
    const history: SpendingHistoryEntry[] = orders.map(order => ({
      orderId: order.id,
      amount: order.totalAmount,
      date: order.orderDate,
      itemCount: order.items.length,
    }));

    return history.sort((a, b) => b.date.getTime() - a.date.getTime());
  }

  public calculateCustomerLifetimeValue(customerId: number): number {
    if (!this.customers.has(customerId)) {
      throw new Error('Customer not found');
    }

    const orders = this.orderHistory.get(customerId) || [];
    return orders.reduce((total, order) => total + order.totalAmount, 0);
  }

  public getCustomerProfile(customerId: number): CustomerProfile {
    const customer = this.customers.get(customerId);
    if (!customer) {
      throw new Error('Customer not found');
    }

    return {
      personal: {
        id: customer.id,
        firstName: customer.firstName,
        lastName: customer.lastName,
        email: customer.email,
        phone: customer.phone || null,
        status: customer.status,
      },
      addresses: this.addresses.get(customerId) || [],
      marketing: this.marketingPreferences.get(customerId) || null,
      orderHistory: this.getCustomerSpendingHistory(customerId),
      lifetimeValue: this.calculateCustomerLifetimeValue(customerId),
    };
  }

  private findCustomerByEmail(email: string): Customer | null {
    for (const customer of this.customers.values()) {
      if (customer.email === email) {
        return customer;
      }
    }
    return null;
  }

  private recordFailedLoginAttempt(email: string): void {
    const attempts = this.loginAttempts.get(email) || [];
    attempts.push(new Date());
    this.loginAttempts.set(email, attempts);
  }

  private isAccountLocked(customerId: number): boolean {
    const customer = this.customers.get(customerId);
    if (!customer) {
      return false;
    }

    const attempts = this.loginAttempts.get(customer.email) || [];
    const fifteenMinutesAgo = new Date(Date.now() - 15 * 60 * 1000);
    const recentAttempts = attempts.filter(
      attempt => attempt > fifteenMinutesAgo
    );

    return recentAttempts.length >= 3;
  }

  private clearFailedLoginAttempts(customerId: number): void {
    const customer = this.customers.get(customerId);
    if (customer) {
      this.loginAttempts.delete(customer.email);
    }
  }

  private updateLastLogin(customerId: number): void {
    const customer = this.customers.get(customerId);
    if (customer) {
      const updatedCustomer: Customer = {
        ...customer,
        lastLogin: new Date(),
      };
      this.customers.set(customerId, updatedCustomer);
    }
  }

  private isValidEmail(email: string): boolean {
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return emailRegex.test(email);
  }

  private isValidPhone(phone: string): boolean {
    const phoneRegex = /^\+?[1-9]\d{1,14}$/;
    return phoneRegex.test(phone);
  }

  private hashPassword(password: string): string {
    // Simple hash simulation for this exercise
    return `hashed_${password}`;
  }

  private verifyPassword(plaintext: string, hashed: string): boolean {
    return hashed === `hashed_${plaintext}`;
  }
}
